dj/api/nodes.py

```diff
@@ -343,7 +340,7 @@ def create_node_revision(
         raise DJException(
             f"Cannot create nodes with multi-catalog dependencies: {set(catalog_ids)}",
         )
-    catalog_id = next(iter(catalog_ids), "draft")
+    catalog_id = next(iter(catalog_ids), 0)
```

Comment:
I was wondering how to handle this in a previous PR and curious to hear your thoughts. When a node is created with no missing parent nodes, everything works nicely and it inherits the catalog. But when a draft node is added with missing parent nodes, there's no way to know the catalog until those parents are added. So it felt like a choice between making catalogs optional on the model or defaulting them to something. `"default"` didn't feel right because that's an actual catalog that we would offer "out of the box" and could get confusing if for some reason when the missing parent nodes get added their catalog value isn't propagated properly. Just thinking out loud here.

Reply:
@samredai I'm confused here, why does a `NodeRevision` have a catalog (and schema/table)? The catalog/schema/table should be a query concept, and not part of the node itself.

Reply:
The catalog is just a logical concept for the nodes so that the "common catalog" validation can be done through change management. That way the build doesn't have to do that validation on the fly. Catalogs can have multiple "engines" though and that's all handled in the query service so it's still a query concept if I understand what you meant by that. For example you can have a postgres, trino, and druid engine defined in the query service that all run the query against different locations.

Reply:
But with this we lost support for source nodes that are materialized in different catalogs, which is a huge feature in the long run. In addition, if we later move the data to a different catalog, we won't be able to compute an older node version in the new catalog, since the version is tied to a specific catalog.

Reply:
I might not fully understand the feature but we've talked about supporting multi-catalog joins in the future, would this feature be part of that? I added this as a topic for the next sync.

Reply:
Oh no, I'm not talking about multi-catalog joins. I'm talking about having multiple catalogs where a given node can be computed.

I'm trying to understand where this change (introduction of revisions?) happened, but originally any node could have multiple representations. In the original examples some source nodes pointed to tables in Druid/Postgres/GSheets, some pointed to tables in in Druid/Postgres, and so on. The query engine would evaluate a request and figure out the best database to fulfill the request. Depending on the columns in the request it would use Druid, or Postgres, or GSheets.

Some requests couldn't be fulfilled, because there was no common database (or catalog, today). For cases like that in the future we could implement multi-catalog joins. But even without implementing the multi-catalog joins **it's important to decouple the concept of a node from where it's materialized (or can be computed). So nodes should not have catalogs, schemas, tables associated with them.**

Reply:
I see, I think this change is from the recent big PR that got merged yesterday (as @shangyian suggested).

> Some requests couldn't be fulfilled, because there was no common database (or catalog, today). For cases like that in the future we could implement multi-catalog joins. But even without implementing the multi-catalog joins it's important to decouple the concept of a node from where it's materialized (or can be computed). So nodes should not have catalogs, schemas, nor tables associated with them.

So if I understand correctly, this complexity lived in the build logic where it had to account for the variable locations where data for a node may actually be available, also considering certain limitations such as data in druid not being able to join to google sheets? Can you give some use cases where the same exact data would exist in different tables in different systems? Are these multiple representations of the data equivalent?

Reply:
> So if I understand correctly, this complexity lived in the build logic where it had to account for the variable locations where data for a node may actually be available, also considering certain limitations such as data in druid not being able to join to google sheets?

Right, when fulfilling a request the query builder would figure out which database could execute the query, taking into consideration where the nodes are materialized and the schemas of the tables. If you had (brackets are databases where the node is materialized):

```
A[db1] B[db1, db2]
 \    /
  C[db3]
```

Then the node `C` could be computed in either `db1` or `db3`, but not `db2` (replace database with catalog for the new design). Furthermore, it also accounted for heterogeneous schemas and the cost of each database/table when doing that evaluation. Eg, in the original examples the dimension node had a `secret_column` that existed only in GSheets (a slow DB). Queries that requested that column would then be evaluated in GSheets, all other would use a faster database.

> Can you give some use cases where the same exact data would exist in different tables in different systems?

At Facebook we had data arriving in Kafka-like streams, and they would be stored in Hive. A single source node would point to both, and we could generate streaming SQL or Presto SQL when computing metrics, using the former if needed more recent data and the latter for historical queries.

There's also the use case where a source node can have **multiple tables** in the same schema. At Facebook we had `dim_all_users` (a user dimension table that had 100+ columns) and `dim_fast_users` (a version with fewer columns, only the ones that could be loaded quickly in the daily job). The `users` node pointed to both, and DJ would use the fastest table if it could, unless it needed columns from the slow table.

> Are these multiple representations of the data equivalent?

Yes, they're conceptually the same, and should be consistent, even if they have different schemas or different time ranges of data.

Reply:
I really like the engine/catalog separation. It makes a better match to real the data stack. But whether the catalog should be part of the node itself or part of the node materialization attributes that's a big question.